Health service

rubenOrtz opened the pull request:
- Native Health service implementation
- Extended health service with per service checker

Commit:
Health Service first look

Implemented native + custom (next commit) health service for spring boot. First implementation. Documentation: https://docs.spring.io/spring-boot/api/rest/actuator/health.html

Commit:
Health System v1

Added healthCheck() for every needed service
Changed errors in order to being able to handle more than one at a time

pabloasenjo3 commented on src/main/java/com/quimify/api/molecularmass/MolecularMassService.java:
```diff
@@ -88,6 +92,21 @@ MolecularMassResult tryCalculate(String query) {
         return molecularMassResult;
     }
 
+    public HealthResult healthCheck() {
```
Move to internal (commented) section of the class, not client section

pabloasenjo3 commented on src/main/java/com/quimify/api/organic/OrganicService.java:
```diff
@@ -98,6 +100,32 @@ OrganicResult tryGetFromStructure(int[] inputSequence) {
+        try {
+            String testCompoundName = "metano";
+            // TODO: Change to a valid compound Structure
+            int[] testCompoundStructure = { 0, 0, 0, 0, 0 };
```
Use {0, 18}

pabloasenjo3 commented on src/main/java/com/quimify/api/health/HealthService.java:
```diff
@@ -0,0 +1,60 @@
+        this.classificatService = classificationService;
+    }
+
+    public HealthResult checkOverallHealth() {
```
Rename "health" as in controller

pabloasenjo3 replied:
Make this method return String

pabloasenjo3 replied:
It should be a series of ifs calling each boolean checkHealth method. When any is false, add a String error to an error list. 

Then return `errorList.toString()`.

If `errorList` is not empty, handle error using [this](https://github.com/quimifyapp/api/blob/97cfa482cf9ccdf6d7776ca81dc9de68d5064370/src/main/java/com/quimify/api/error/ErrorService.java#L27)

rubenOrtz replied:
The errors list is already returned in the method getErrors() at HealthResult. Added the ErrorService handler.

pabloasenjo3 commented on src/main/java/com/quimify/api/inorganic/InorganicService.java:
```diff
@@ -79,6 +81,21 @@ InorganicResult search(String input) {
         return deepSearch(input);
     }
 
+    public HealthResult healthCheck() {
```
Rename checkHealth and make it boolean, same with others.
Result classes are typically for REST responses

pabloasenjo3 commented on src/main/java/com/quimify/api/health/HealthController.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    @GetMapping
+    public Health getHealth() {
```
Simply returning String is OK for this matter

pabloasenjo3 commented on src/main/java/com/quimify/api/health/HealthController.java:
```diff
@@ -0,0 +1,22 @@
+
+    private final QuimifyHealthIndicator healthIndicator;
+
+    public HealthController(QuimifyHealthIndicator healthIndicator) {
```
There is no need for this indicator. Simply inject a service, as done [here](https://github.com/quimifyapp/api/blob/97cfa482cf9ccdf6d7776ca81dc9de68d5064370/src/main/java/com/quimify/api/client/ClientController.java#L17)

pabloasenjo3 commented on src/main/java/com/quimify/api/health/HealthController.java:
```diff
@@ -0,0 +1,22 @@
+
+    @GetMapping
+    public Health getHealth() {
+        return healthIndicator.health();
```
Add a log as done [here](https://github.com/quimifyapp/api/blob/97cfa482cf9ccdf6d7776ca81dc9de68d5064370/src/main/java/com/quimify/api/client/ClientController.java#L27)

pabloasenjo3 commented on src/main/java/com/quimify/api/health/QuimifyHealthIndicator.java:
```diff
@@ -0,0 +1,35 @@
+        HealthResult healthResult = healthService.checkOverallHealth();
+
+        if (healthResult.isPresent()) {
+            return new Health.Builder()
```
This logic belongs in `@Service`

rubenOrtz replied:
This cannot be done in an @Service because it is an implementation of the HealthIndicator interface provided by the Spring Framework.

pabloasenjo3 replied:
Just delete `QuimifyHealthIndicator.java`

Commit:
Changes based on pull request

pabloasenjo3 commented on src/main/java/com/quimify/api/health/HealthResult.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class HealthResult {
```
Just `boolean healthy` plus `List<String> errors`

rubenOrtz replied:
Okey! Now only uses this.healthy, this.errors & this.message.
This.message is still needed In order to show the message: "System errors detected: " + errors

pabloasenjo3 replied:
Without `String message` is OK

pabloasenjo3 commented on src/main/java/com/quimify/api/health/HealthService.java:
```diff
@@ -0,0 +1,61 @@
+    @Autowired
+    ErrorService errorService;
+
+    public HealthService(InorganicService inorganicService, OrganicService organicService,
```
Just use `@Autowired` like with `
@Autowired
ErrorService errorService;`.
There is no need for a constructor

rubenOrtz replied:
You're absolutely right.

pabloasenjo3 commented on src/main/java/com/quimify/api/health/HealthController.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    @GetMapping
+    public Health getHealth() {
```
If you don't want to return `String`, then return `HealthResult` instead. Use `Health` only in `HealthService`.
Classes like `HealthResult` are meant to be sent to the client, and are not meant to be used internally

rubenOrtz replied:
Okey! In addition, I also used @Autowired for HealthService.

pabloasenjo3 commented on src/main/java/com/quimify/api/health/HealthService.java:
```diff
@@ -0,0 +1,61 @@
+    public final ClassificationService classificatService;
+
+    @Autowired
+    ErrorService errorService;
```
Unused code

pabloasenjo3 commented on src/main/java/com/quimify/api/health/HealthService.java:
```diff
@@ -0,0 +1,61 @@
+        if (!subSystemHealth.isPresent()) {
+            overallHealth.setPresent(false);
+            overallHealth.addError(subSystemHealth.getMessage());
+            ;
```
Disregarded code!

pabloasenjo3 replied:
Be careful and always pay attention to detail

pabloasenjo3 commented on src/main/java/com/quimify/api/classification/ClassificationService.java:
```diff
@@ -46,6 +48,24 @@ public boolean isInorganic(Classification classification) {
+            Optional<Classification> inorganicFormulaResult = classify(testFormula);
+
+            if (inorganicFormulaResult.isEmpty() || inorganicFormulaResult.get() != Classification.inorganicFormula) {
+                throw new RuntimeException("Error classifying inorganic formula: " + testFormula);
```
Why throwing an Exception?
Just return appropiate value (`false`)

pabloasenjo3 commented on src/main/java/com/quimify/api/molecularmass/MolecularMassService.java:
```diff
@@ -62,6 +66,22 @@ public Optional<Float> get(String formula) { // TODO just use tryCalculate?
+            String testFormula = "H"; // Expected ~1.00794
+            MolecularMassResult result1 = tryCalculate(testFormula);
+            if (!result1.isPresent()) {
+                throw new RuntimeException("Error calculating molecular mass: " + testFormula);
```
Same thing

pabloasenjo3 commented on src/main/java/com/quimify/api/organic/OrganicService.java:
```diff
@@ -53,6 +55,34 @@ class OrganicService {
+            // 1. Check name-to-structure resolution
+            Optional<OrganicResult> resultFromName = tryGetFromName(testCompoundName);
+            if (resultFromName.isEmpty() || !resultFromName.get().isFound()) {
+                throw new RuntimeException("Error resolving organic compound from name: " + testCompoundName);
```
Same thing

pabloasenjo3 commented on src/main/java/com/quimify/api/inorganic/InorganicService.java:
```diff
@@ -55,6 +57,23 @@ class InorganicService {
+        try {
+            Optional<Integer> cachedId = cacheComponent.find(Normalizer.get("h2o"));
+            if (cachedId.isEmpty()) {
+                throw new RuntimeException("Error al acceder al caché de inorgánicos");
```
Same thing

pabloasenjo3 reviewed (changes requested):
Added new issues. Also marked some old issues as **unresolved**

Commit:
Changes according to the pull request.

Commit:
Deleted QuimifyHealthIndicator

pabloasenjo3 commented on src/main/java/com/quimify/api/health/HealthResult.java:
```diff
@@ -4,18 +4,18 @@
     private String message;
     private List<String> errors;
 
-    public HealthResult(boolean present, String message) {
```
Use two (three) constructors:
1. Public one with a list parameter
2. Named constructor healthy()
3. Private constructor for number 2 to work

pabloasenjo3 commented on src/main/java/com/quimify/api/health/HealthResult.java:
```diff
@@ -24,13 +24,13 @@ public void addError(String error) {
     }
 
     // Getters
```
Make them trivial as done [here](https://github.com/quimifyapp/api/blob/e394fe2bb3920309db4d6dbfa011c8ec67123cd2/src/main/java/com/quimify/api/inorganic/InorganicResult.java#L98)

pabloasenjo3 commented on src/main/java/com/quimify/api/health/HealthService.java:
```diff
@@ -14,21 +13,17 @@
+    public ClassificationService classificatService;
 
     public HealthResult health() {
         HealthResult overallHealth = new HealthResult(false, ""); // Start pessimistic, no errors yet
```
Declare an errors list here instead

pabloasenjo3 replied:
Then create a `HealthResult` just when it's time to return

Commit:
Updated logic in order to follow the best practices

Key Changes:

Introduced HealthCheck Interface: Created an interface to decouple health check logic from the HealthService and HealthResult classes.

Refactored Services: Updated all services (InorganicService, etc.) to implement the HealthCheck interface and return an error message or null.

Leveraged Dependency Injection: In HealthService, injected a list of all HealthCheck implementations to automatically discover and execute health checks.
(Every component that want to be checked need a checkHealth and implements HealthCheck)

Simplified HealthResult: Removed unnecessary getMessage(), setHealthy() methods and message logic.

Improved HealthController: Now returns a ResponseEntity with appropriate HTTP status codes (200 OK or 500 Internal Server Error) based on the health check results.

Benefits:

Decoupled Design: Individual services are independent of the core health check logic.

Enhanced Maintainability: Easier to modify the health check process without impacting individual services.

Streamlined Testing: Simplified unit testing of each service's health check independently.

Cleaner & Easier Code.

Commit:
A detail

pabloasenjo3 committed:
Make service health checks boolean and re-couple code
